### ISSUE TITLE
Path Status is ill defined

### ISSUE THREAD

**Issue by huitema** [quicwg/multipath#218]
The definition of the path status frame is somewhat ambiguous. When trying to implement that, I have many questions:

1) If paths have a status, what is the default status of a path just after validation, and before receiving any Status frame?

2) Who send this? From the draft, I think "either endpoint". But then, what does that mean if one endpoint says "available" and the other says "standby"?

3) If a path is standby state and the endpoint receives validating frames on that path, does the status change?

4) If a path is in standby state and no other path is in available state, does the standby path automatically become available?

5) Do we really need to maintain a sequence counter for status frames per path and per direction?

**Comment by qdeconinck** [comment]
From the IETF 116 minutes (https://github.com/quicwg/wg-materials/blob/main/ietf116/minutes.md#multipath), there were support in splitting the current PATH_STATUS frame into smaller ones to avoid handling invalid/unknown values (e.g., PATH_AVAILABLE and a PATH_STANDBY frames).

Still, the points raised above would also apply on such splitter frames. From my current understanding,

> If paths have a status, what is the default status of a path just after validation, and before receiving any Status frame?

As long as no status information is received, I would consider a path as being "available" for use.

> Who send this? From the draft, I think "either endpoint". But then, what does that mean if one endpoint says "available" and the other says "standby"?

That's a good point. I see use cases from a client-sending perspective, but I don't see why a server could not send statuses as well. The PATH_STATUS frame gives an indication about what should be its peer's path scheduling strategy, but it has no way of forcing the other to do so...

> If a path is standby state and the endpoint receives validating frames on that path, does the status change?

I would decouple the validation state of a network path from its status usage. In other words, I would not use implicit signals to change the (usage) status requested by the peer over a path.

> If a path is in standby state and no other path is in available state, does the standby path automatically become available?

I would say that if your peer advertises the path as being standby, you SHOULD not use it for non-probing packets, unless there is no working available paths. The concrete implementation of such a behaviour is implementation-specific, as this is a scheduling issue (starting from when do you consider that a path is not working anymore? if it experiences RTOs or being abandoned/closed?). Furthermore, in the case you have several standby paths (and no available ones), it is up to the endpoint to decide which path(s) to use (until receiving explicit signals).

> Do we really need to maintain a sequence counter for status frames per path and per direction?

For sending, I think you could use a single counter for all paths and directions, if wanted. But at receiving side, each path needs to track the highest sequence number that updated its status, to cope with reordered receive patterns. For instance, if you have p1 being standby by the seq being 19, and p2 being available, updated by seq 18, and the receive frames in the following order:

- Status (p1, seq=22, available) : highest seq on p1, update p1 to available
- Status (p2, seq=21, standby): highest seq on p2, update p2 to standby
- Status (p1, seq=20, standby): not the highest seq on p1, ignore old information

**Comment by mirjak** [comment]
> > If paths have a status, what is the default status of a path just after validation, and before receiving any Status frame?
> 
> As long as no status information is received, I would consider a path as being "available" for use.

I think you should consider it as available for use but I don't think you necessarily should set the status. The status is simply that you didn't receive any status information.

> 
> > Who send this? From the draft, I think "either endpoint". But then, what does that mean if one endpoint says "available" and the other says "standby"?
> 
> That's a good point. I see use cases from a client-sending perspective, but I don't see why a server could not send statuses as well. The PATH_STATUS frame gives an indication about what should be its peer's path scheduling strategy, but it has no way of forcing the other to do so...

Actually we say "An endpoint uses PATH_STATUS frames to inform that the peer should send packets in the preference expressed by these frames."

That means it always the receiver that informs the sender. Note that as we don't have a path identified anymore the PATH_STATUS information in unidirectional and therefore it is possible that one peer says it okay to send on a certain path, while the other end does not what to receive anything on that same path.

> 
> > If a path is standby state and the endpoint receives validating frames on that path, does the status change?

No. The path status is only a recommendation but it doesn't forbid sending: "PATH_STATUS frame to express a preference in path usage". However, I guess that not fully clear in the frame definition section as we use normative SHOULDs there. Seem like an editorial clarification could make sense.

> 
> I would decouple the validation state of a network path from its status usage. In other words, I would not use implicit signals to change the (usage) status requested by the peer over a path.
> 
> > If a path is in standby state and no other path is in available state, does the standby path automatically become available?

No. But it can be used as it is just a recommendation. See also discussion in issue #238.

> 
> I would say that if your peer advertises the path as being standby, you SHOULD not use it for non-probing packets, unless there is no working available paths. The concrete implementation of such a behaviour is implementation-specific, as this is a scheduling issue (starting from when do you consider that a path is not working anymore? if it experiences RTOs or being abandoned/closed?). Furthermore, in the case you have several standby paths (and no available ones), it is up to the endpoint to decide which path(s) to use (until receiving explicit signals).
> 
> > Do we really need to maintain a sequence counter for status frames per path and per direction?

The sequence number is per connection but you have to hold state per path anyway as Quentin described below.

> 
> For sending, I think you could use a single counter for all paths and directions, if wanted. But at receiving side, each path needs to track the highest sequence number that updated its status, to cope with reordered receive patterns. For instance, if you have p1 being standby by the seq being 19, and p2 being available, updated by seq 18, and the receive frames in the following order:
> 
> * Status (p1, seq=22, available) : highest seq on p1, update p1 to available
> * Status (p2, seq=21, standby): highest seq on p2, update p2 to standby
> * Status (p1, seq=20, standby): not the highest seq on p1, ignore old information

Do we now need any text updates @huitema ?

**Comment by kazuho** [comment]
@mirjak 
> The sequence number is per connection but you have to hold state per path anyway as Quentin described below.

I think I would prefer changing this sequence number to be a per-DCID counter rather than a per-connection counter.

The rationale is that the path status is a per-DCID variable therefore it is natural to have a per-DCID sequence number; it is counterintuive to use a per-connection sequence number for communicating a per-DCID state.

Moreover, using a per-connection counter requires more state than using per-DCID sequence number.

Regardless of the sequence number being per-connection or per-DCID, the sender and receiver has to remember the up-to-date sequence number for each DCID. Without such information, the sender cannot determine if it should retransmit when a packet is deemed lost; the receiver cannot determine if the received frame contains the newest information.

In addition, by defining the sequence number as per-connection, the specification is requiring the sender to retain another state which is the per-connection sequence number. If we change the sequence number to be per-DCID, we do not need such state.

To summarize, using a per-DCID sequence number is more natural (because each state being sent will have it's own sequence number) and requires less state.

**Comment by qdeconinck** [comment]
I agree with Kazuho here in the sense that we should not enforce the "path status sequence number" to be monotonically increasing over the connection. It just needs to be monotonically increasing over each DCID to let the receiver reorder the path statuses properly. Then, it is up to the implementation whether to keep a global monotonically increasing sequence number at sending side or a local one per DCID -- but the specification should not enforce the first solution.

**Comment by mirjak** [comment]
@huitema I think we can close this issue with the merge of PR #242 ?

**Comment by huitema** [comment]
Yes, we can close this issue.
